### PR TITLE
docs: add migration guide for tsdb

### PR DESCRIPTION
**Pull request by ashwanthgoli**
**What this PR does / why we need it**:
Adds a migration guide for moving from any of the older indexes to TSDB

**Special notes for your reviewer**:

**Checklist**
- [x] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [x] Documentation added
- [ ] Tests updated
- [ ] `CHANGELOG.md` updated
  - [ ] If the change is worth mentioning in the release notes, add `add-to-release-notes` label
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md`
- [ ] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)

### CODE REVIEW

File: docs/sources/setup/migrate/migrate-to-tsdb/_index.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+This example adds a new `period_config` which configures Loki to start using tsdb index for the data ingested from `2023-10-20`.
+
+①  Make sure you are setting the new period `from` to a date in the future.
+
+②  Update the new period to use **tsdb** as the index type.
+
+③  This setup uses filesystem as the storage in both the periods. But if you desire to store the `tsdb` index along with it's chunks in a differnt object store, you can update the `object_store` in the new period.
+
+④  Optionally update the schema to v12 which is the recommended verison at the time of writing. Please refer to [configure page]({{< relref "../../../configure#period_config" >}}) for the current recommend version.
```
Comment (salvacorts) [inline docs/sources/setup/migrate/migrate-to-tsdb/_index.md:50]:
```suggestion
④  Optionally update the schema to v12 which is the recommended version at the time of writing. Please refer to [configure page]({{< relref "../../../configure#period_config" >}}) for the current recommend version.
```

Comment (salvacorts) [inline docs/sources/setup/migrate/migrate-to-tsdb/_index.md:50]:
```suggestion
④  Optionally update the schema to v12 which is the recommended verison at the time of writing. Please refer to the [configure page]({{< relref "../../../configure#period_config" >}}) for the current recommend version.
```

File: docs/sources/setup/migrate/migrate-to-tsdb/_index.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+{{% admonition type="note" %}}
+It's important that you rollout the new `period_config` change to all loki components for it to take effect.
```
Comment (salvacorts) [inline docs/sources/setup/migrate/migrate-to-tsdb/_index.md:53]: This is pretty important, I'd place this note before the code snippet.

File: docs/sources/setup/migrate/migrate-to-tsdb/_index.md [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+{{% /admonition %}}
+
+### Configure TSDB shipper
+
+It's also important that you configure the `tsdb_shipper` block in [storage_config]({{< relref "../../../configure#storage_config" >}}). Mainly the following options:
+- `active_index_directory`: directory where ingesters would write index files which would then be uploaded by shipper to configured storage.
+- `cache_location`: cache location for downloading index files from the storage for use in query path.
+
+```
+tsdb_shipper:
```
Comment (salvacorts) [inline docs/sources/setup/migrate/migrate-to-tsdb/_index.md:63]: This should go inside `storage_config` right?

File: docs/sources/setup/migrate/migrate-to-tsdb/_index.md [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+  active_index_directory: /data/tsdb-index
+  cache_location: /data/tsdb-cache
+```
+
+### Run compactor
+
+It is strongly recommended that you run compactor when using TSDB index. It is responsible for running compaction and retention on TSDB index.
```
Comment (salvacorts) [inline docs/sources/setup/migrate/migrate-to-tsdb/_index.md:70]:
```suggestion
We strongly recommended running the compactor <add link?> when using TSDB index. It is responsible for running compaction and retention on TSDB index.
```

**Review by salvacorts — APPROVED** [review]
LGTM. left some minor comments

File: docs/sources/setup/migrate/migrate-to-tsdb/_index.md [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+
+### Configure TSDB index for an upcoming period
+
+To begin the migration, add a new [period_config]({{< relref "../../../configure#period_config" >}}) entry in your [schema_config]({{< relref "../../../configure#schema_config" >}}).
+You can read more about schema config [here]({{< relref "../../../storage#schema-config" >}}).
+
+{{% admonition type="note" %}}
+It's important that you rollout the new `period_config` change to all loki components for it to take effect.
+{{% /admonition %}}
```
Comment (JStickler) [inline docs/sources/setup/migrate/migrate-to-tsdb/_index.md:25]:
```suggestion
{{% admonition type="note" %}}
You must roll out the new `period_config` change to all Loki components in order for it to take effect.
{{% /admonition %}}
```

File: docs/sources/setup/migrate/migrate-to-tsdb/_index.md [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    - from: 2023-10-20 ①
+      store: tsdb ②
+      object_store: filesystem ③
+      schema: v12 ④
+      index:
+        prefix: index_
+        period: 24h
+```
+
+This example adds a new `period_config` which configures Loki to start using tsdb index for the data ingested from `2023-10-20`.
```
Comment (JStickler) [inline docs/sources/setup/migrate/migrate-to-tsdb/_index.md:46]: Consider moving this line above the example.
```suggestion
This example adds a new `period_config` which configures Loki to start using the TSDB index for the data ingested starting from `2023-10-20`.
```

File: docs/sources/setup/migrate/migrate-to-tsdb/_index.md [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+we strongly recommend migrating to TSDB.
+
+
+### Configure TSDB index for an upcoming period
+
+To begin the migration, add a new [period_config]({{< relref "../../../configure#period_config" >}}) entry in your [schema_config]({{< relref "../../../configure#schema_config" >}}).
+You can read more about schema config [here]({{< relref "../../../storage#schema-config" >}}).
+
+{{% admonition type="note" %}}
+It's important that you rollout the new `period_config` change to all loki components for it to take effect.
```
Comment (JStickler) [inline docs/sources/setup/migrate/migrate-to-tsdb/_index.md:24]:
```suggestion
You must roll out the new `period_config` change to all Loki components in order for it to take effect.
```

File: docs/sources/setup/migrate/migrate-to-tsdb/_index.md (hunk 6 above)
Comment (JStickler) [inline docs/sources/setup/migrate/migrate-to-tsdb/_index.md:46]: Consider moving this line above the example code.
```suggestion
This example adds a new `period_config` which configures Loki to start using the TSDB index for the data ingested starting from `2023-10-20`.
```

File: docs/sources/setup/migrate/migrate-to-tsdb/_index.md [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+      index:
+        prefix: index_
+        period: 24h
+```
+
+This example adds a new `period_config` which configures Loki to start using tsdb index for the data ingested from `2023-10-20`.
+
+①  Make sure you are setting the new period `from` to a date in the future.
+
+②  Update the new period to use **tsdb** as the index type.
```
Comment (JStickler) [inline docs/sources/setup/migrate/migrate-to-tsdb/_index.md:50]:
```suggestion
②  Update the new `from` period to use TSDB as the index type by setting `store: tsdb`.
```

File: docs/sources/setup/migrate/migrate-to-tsdb/_index.md [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+      object_store: filesystem ③
+      schema: v12 ④
+      index:
+        prefix: index_
+        period: 24h
+```
+
+This example adds a new `period_config` which configures Loki to start using tsdb index for the data ingested from `2023-10-20`.
+
+①  Make sure you are setting the new period `from` to a date in the future.
```
Comment (JStickler) [inline docs/sources/setup/migrate/migrate-to-tsdb/_index.md:48]:
```suggestion
①  You must set the new `from` period to a date in the future.
```

File: docs/sources/setup/migrate/migrate-to-tsdb/_index.md [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+
+②  Update the new period to use **tsdb** as the index type.
+
+③  This setup uses filesystem as the storage in both the periods. But if you desire to store the `tsdb` index along with it's chunks in a differnt object store, you can update the `object_store` in the new period.
```
Comment (JStickler) [inline docs/sources/setup/migrate/migrate-to-tsdb/_index.md:52]:
```suggestion
③  This sample uses filesystem as the storage in both the `from` periods. If you want to use a different storage for the TSDB index and chunks, you can specify a different `object_store` when you define the new `from` period.
```

File: docs/sources/setup/migrate/migrate-to-tsdb/_index.md [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+
+④  Optionally update the schema to v12 which is the recommended version at the time of writing. Please refer to the [configure page]({{< relref "../../../configure#period_config" >}}) for the current recommend version.
```
Comment (JStickler) [inline docs/sources/setup/migrate/migrate-to-tsdb/_index.md:54]: If this is the recommended schema, let's just tell them to update it.
```suggestion
④  Update the schema to v12 which is the recommended version at the time of writing. Please refer to the [configure page]({{< relref "../../../configure#period_config" >}}) for the current recommend version.
```

File: docs/sources/setup/migrate/migrate-to-tsdb/_index.md [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+
+### Configure TSDB shipper
+
+It's also important that you configure the `tsdb_shipper` block in [storage_config]({{< relref "../../../configure#storage_config" >}}). Mainly the following options:
```
Comment (JStickler) [inline docs/sources/setup/migrate/migrate-to-tsdb/_index.md:58]:
```suggestion
It's also important that you configure the `tsdb_shipper` block in [storage_config]({{< relref "../../../configure#storage_config" >}}). Specifically the following options:
```

File: docs/sources/setup/migrate/migrate-to-tsdb/_index.md [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+- `active_index_directory`: directory where ingesters would write index files which would then be uploaded by shipper to configured storage.
```
Comment (JStickler) [inline docs/sources/setup/migrate/migrate-to-tsdb/_index.md:59]:
```suggestion
- `active_index_directory`: directory where ingesters will write index files which would then be uploaded by shipper to configured storage.
```

File: docs/sources/setup/migrate/migrate-to-tsdb/_index.md [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+storage_config:
+  tsdb_shipper:
+    active_index_directory: /data/tsdb-index
+    cache_location: /data/tsdb-cache
+```
+
+### Run compactor
+
+We strongly recommended running the [compactor]({{< relref "../../../operations/storage/retention#compactor" >}}) when using TSDB index. It is responsible for running compaction and retention on TSDB index.
+Not running index compaction would result in sub-optimal query performance.
```
Comment (JStickler) [inline docs/sources/setup/migrate/migrate-to-tsdb/_index.md:72]:
```suggestion
Not running index compaction will result in sub-optimal query performance.
```

**Review by JStickler — COMMENTED** [review]
[docs team] Looking good.  Thanks for writing this up, it will be really helpful to get users onto TSDB.

**Comment by ashwanthgoli** [comment]
@JStickler i am inclined to leave it as period, or we could update it to `period_config` to be explicit

**Comment by JStickler** [comment]
> i am inclined to leave it as period, or we could update it to period_config to be explicit

@ashwanthgoli I'm fine with just saying "period", I couldn't think of a better way to refer to that chunk of the configuration file.

**Pull request merged**